rfcs/text/0428-extending-and-duplicating-types.md

```diff
@@ -0,0 +1,991 @@
+  ... // superpower, age, occupation
+}
+```
+
+And then we instead try and create one with an array of `name`s:
+
+```json5
+{
+  name: ["John Doe"], // ERROR: expected string, got string array
+  ... // superpower, age, occupation
+}
+```
+
+### An uncomfortable compromise
+
+As alluded to above, checking compatibility of any two given schemas is quite possibly an NP-Hard problem. Even if it is generally solvable, it is likely a task that is complicated enough that it would be an exceptionally high barrier to entry for potential implementers to overcome.
+
+While various implementations likely will want to implement checking for the simpler cases (disallowing a property from being two disjoint types for example), we opt not to encode such restrictions within the type system, and **implementations should not assume that an arbitrary type is satisfiable**.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+- [JSON Schema composition](https://json-schema.org/understanding-json-schema/reference/combining.html) and [`unevaluatedProperties`](https://json-schema.org/understanding-json-schema/reference/object.html#unevaluated-properties)
+- [Programming languages subtyping](https://en.wikipedia.org/wiki/Subtyping)
+- [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- We haven't specified how projecting/selecting properties of a supertype from a subtype instance is possible. It is an open question how we actually pick out the exact properties of a subtype to provide a valid supertype instance in embedding applications.
+  - This may be especially difficult if a property type is defined as being an array of `oneOf` a given set of property type objects. Detecting which sub-schema applies to a given array element when accounting for dropped properties, **may** be a very difficult task.
+- Type extension might prove to have strange side-effects when considering link types which extend from one another.
+  - Say `Person` defines `Knows` and `Has Friend` links. If `Has Friend` extends `Knows`, then technically any `Has Friend` instance is also a `Knows` instance, and would need to satisfy the link constraints set on both. This has implications for `minItems` and `maxItems`, as well as for link destinations. If `Person` says it only `Knows` other `Person` entities, but `Has Friend` links to both `Person` and `Pet` entities, then a `Has Friend` link to a `Pet` entity would break the constraint set for `Knows`.
+  - Say `Person` defines `Knows` and `Has Friend` links, and that those links are **ordered**. At the moment, the order of a link is defined in a given direction (e.g. left to right) and has a single value. That would presently imply that the order of the `Knows` link _within the `Knows` links list_ would have to be the same as the order of the `Has Friend` link _within the `Has Friend` links list_.
```

Comment:
This is definitely a problem, and probably blocks acceptance. 

One potential route is to somehow 'bucket' the lists of links, and have an order per list.

Another potential route is to rework our ordering strategy where we define an `orderBy` mechanism on the source entity type, rather than having a native definition of `order` _within_ link entities. Something like this was proposed by @indietyp a while ago, and has sat in the background until we needed to revisit link ordering (thinking about race-conditions, fractional indexing, consistency guarantees, etc.), now is probably the time where we need to revisit it.

cc: @CiaranMn

Reply:
OrderBy would work, if "user-specified custom" were an option it allowed.

But that option needing to exist would then again necessitate a solution like buckets.

Reply:
> OrderBy would work, if "user-specified custom" were an option it allowed.

Yes I guess the main problem of something like `OrderBy` is it precludes people from designing something like a Spotify playlist. You could easily do things like alphabetical or chronological ordering (which you can't do currently easily), but allowing people to reorder and specify a defined ordering.. that would be local to the playlist, and would have the same issue outlined above about a link appearing under multiple 'collections'.

A potential route away from this, is that we provide an `orderBy` mechanism on links but then also establish a second pattern wherein entities store local definitions of orderings within their properties. So, a `Playlist` entity could store an optional `Song Ordering` property, which is an array of link entity IDs. I _**really**_ don't like this though, and the main reasoning for that is explained within #1020 and is to do with "referability" and consistency guarantees. If a different block / user / whatever were to delete one of the link entities, or somehow make them invalid, the `Playlist` entity would _not_ be implicitly updated because there would be nothing within the type system to enforce it (like there _**is**_ with actual links), and this causes a massive risk vector for race conditions.

On the other hand, one of the issues with `OrderBy` (and why we didn't continue exploring it in detail), is that really that's something a user can do at point of querying, `queryEntities` for instance already allows you to define an `OrderBy` constraint, which could affect pagination. The main benefit would be allowing backends to optimise their storage layout for ordering in the first place and have faster queries, but considering it doesn't solve the core functionality problems here.. we still need to think about the `Playlist` case.

> necessitate a solution like buckets.

I'm hesistant to argue for buckets / collections, but I've brought it up offhand in conversations multiple times over the past few months so it seems like it keeps coming up as a prominent 'obvious' solution to certain things, and maybe it's time to accept it as worth the investment. The investment isn't minor, existing datastores would have to quite heavily change how they store links now (somehow grouping them together which is not a concept we currently have at all), the type system and its implementations would need updating, the query mechanisms would need to change, the `Subgraph` would somehow need to have a way of grouping link entities (tying them to the source entity type definition too which is a bit strange), the standard library would need to change..

Hm

Reply:
I have 3 distinct ideas, that have been bouncing around my head over the last couple of hours. Thought I'd contribute them to the discussion. :D

## link type `orderBy`: Value

We reserve a specific property type (e.g. `blockprotocol/@blockprotocol/orderBy`), if that property type is present in the definition it is ordered. This is not ideal, as we would lose quite a lot of flexibility.

## link type `orderBy`: Path

What if `orderBy` would be defined on the link type itself, but instead of being a value, it would be a (JSON)-pointer to an existing property on the object itself.

The property needs to be a comparable value (preferably a `Number`). The interesting part about this is that it allows the user to choose which property to use freely and can be (pretty efficiently) stored in data stores (by just computing the value).

What would be more complicated is to allow traversal via the JSON pointer (kind of like what we're currently doing with the)

## entity type `orderBy`: Path

instead of defining a specific value, we define a path to a property, the property it is pointing to needs to be of a valid datatype (preferably `Number`). This is very similar to the second proposal, except that we're now able to query the link `rightEntity` (and `leftEntity`) safely, if we navigate to the `rightEntity`, we can only use properties to order that are present on all valid destinations.

### Example

```
EntityType Playlist {
	properties = {
		title,
		description
	}

	links = {
		includedInPlaylist = {
			orderBy = ["properties", "position"]
			// as an alternative
			orderBy = ["rightEntity", "properties", "chartPosition"],
			oneOf = [Song]
		}
	}
}

EntityType Song {
	properties = {
		title,
		description,
		chartPosition,
	}
}

EntityType includedInPlaylist {
	oneOf = [link],

    properties = {
		position
	}
}
```

(the syntax is completely made up, needed something terse as I didn't want to post a whole schema)

What is interesting about this is that by just changing the ordering in `Playlist` we would be able to create custom "auto-generated" playlist positions, that is depending on either the `Song` or the "local" position.

(Of course, the type system would need to have the ability to re-define the `orderBy` attribute)

Reply:
Thanks! To get us all on the same page, I have constructed a (realistic) example that would make exhibit the problem,

let's say we have the following entity types:
* `Person`
* `Address`
* `WorkAddress`
* `HomeAddress`
* `HomeOfficeAddress`
* `Location`

`Person` has 4 links:
* `Person --Address-> Location (orderBy: preference)`
* `Person --WorkAddress-> Location (orderBy: preference)`
* `Person --HomeAddress-> Location (orderBy: preference)`
* `Person --HomeOfficeAddress-> Location (orderBy: preference)`

`Address` defines a property: `preference`, and we have the following inheritance scheme:

```
        Address
       /       \
WorkAddress    HomeAddress
       \       /
    HomeOfficeAddress
```

As an example let's say we have as a user `Max Mustermann`, `Max Mustermann` being the rich banker that he is, is able to afford multiple homes and work locations, doing so we have:

```mermaid
flowchart LR
	mm[Max Mustermann] -->|WorkAddress - 0| Frankfurt
    mm -->|WorkAddress - 10| Berlin
    mm -->|HomeAddress - 20| Hochheim
    mm -->|HomeAddress - 30| Dortmund
    mm -->|HomeOfficeAddress - 40| Ibiza
```

We can generalize these links (and their order property) has part of set theory like this:

<img width="869" alt="image" src="https://user-images.githubusercontent.com/7252775/229898323-af0a682d-ce04-4260-8599-d1a98ddc3b62.png">

Now `Max Musterman` has decided that he prefers `Dortmund` as his `HomeAddress`, so he puts the `preference` to `15`. Now we have also changed the ordering of `Address`, which is unintended (only because I changed my preference for where I live, I don't want to change the preference where someone contacts me.)

To summarize, we have a partially ordered set and, inherited types are essentially subsets into the set. When we reorder in that subset (with the same key) we cannot guarantee that we change the order in the superset.

---

So what if we say: you can use `orderBy`, but if you do, it needs to point to a property defined on the current entity type and cannot be inherited. This limits reuse, but ensures that the subset problem described above does not happen. (if I order, I need to order by something different).

Reply:
I think we should avoid any form of restrictions like "it can't be inherited" as that requires us to resolve the full inheritance tree, and as expressed elsewhere in the RFC, implementations would need to be careful about handling external types which don't abide by the spec (potentially accidentally)

Reply:
didn't really get what 'buckets' is from the thread, is it this? being explicit about which type in the link's type chain an order is in relation to, e.g.
```json
{
  "$id": "https://example.com/types/entity-type/work-address/v/1",
  "leftEntityId": "Max",
  "rightEntityId": "Dortmund",
  "leftToRightOrder": {
     "https://example.com/types/entity-type/work-address/v/1": 15,
     "https://example.com/types/entity-type/address/v/1": 30
  }
}
```
you would need to figure out what to do for types that weren't present in this object and whether they inherited a position or not, and whoever was creating these things would need to figure out from context which order(s) should be set

Reply:
Interesting suggestion as well. I could see this being a royal pain to validate though..

It adds a dependency of the _link entity_ onto the _source entity type_ which isn't ideal. If the source entity was updated (or had its entity type updated) then you could invalidate outgoing link entities.. but I guess as part of validation we'd need to check outgoing link entities anyway as some of them may be required.

> you would need to figure out what to do for types that weren't present in this object 

I personally would treat that as a validation error for the link.

Reply:
At the moment I think that's the best solution of the ones we've come up with

Reply:
> didn't really get what 'buckets' is from the thread, is it this

Apologies @CiaranMn I actually missed that this was a question, I thought it was an alternative proposal. The "buckets" thing I was saying would be on the source entity type, and rather than something like this:
```json5
{
  ... 
  "links": {
    "http...some-link-type": {
        "type": "array",
        "items": {
          "oneOf": [
            {
              "$ref": "http...some-entity-type"
            }
          ]
        }
      },
      ... // more link types
    }
  }
}
```

You'd maybe have something like this

```json5
{
  ... 
  "links": {
    "collection1": { // or 'bucket' 1
        "type": "array",
        "associatedLinkType": "http...some-link-type", // could also perhaps be a `oneOf` if we wanted mixed collections for some reason
        "items": {
          "oneOf": [
            {
              "$ref": "http...some-entity-type" 
            }
          ]
        }
      },
      ... // more link types
    }
  }
}
```

Which means you could do

```json5
{
  ... 
  "links": {
    "collection1": { // or 'bucket' 1
        "type": "array",
        "associatedLinkType": "http...some-link-type",
        "items": {
          "oneOf": [
            {
              "$ref": "http...some-entity-type"
            }
          ]
        }
      },
      ... // more link types
    },
    "collection2": { // or 'bucket' 1
        "type": "array",
        "associatedLinkType": "http...some-link-type", // same link type
        "items": {
          "oneOf": [
            {
              "$ref": "http...some-entity-type" // could be some _other_ entity type (different to above)
            }
          ]
        }
      },
      ... // more link types
    }
  }
}
```

And then when you create a link, it goes in a specific bucket/collection, rather than just being grouped by link _type_. (Therefore having a unique order within that collection).

I'm pretty sure I don't like this suggestion though, it's adding yet another high-level concept, and this one is weirdly implicitly defined within the entity type, rather than being a standalone re-usable thing. I think if you wanted to group links, we'd want it done within the type system and ontology by creating `Group` or `Path` semantics as enabled by the `links/entities` proposal.

Reply:
Basically removing  (currently undefined but implied) guarantees around ordering and putting that onus on the user

Reply:
Thank you for explaining buckets and grouping. Would need to talk about grouping more (probably easier sync) to have thoughts on it.

> > you would need to figure out what to do for types that weren't present in this object

> I personally would treat that as a validation error for the link.

It depends on whether each link was declared as `ordered` on the entity or not, and how we handle validating that generally. If a link was declared `ordered: true` and we treated that as meaning that an ordering property was required (in which direction? either?), I agree we could treat a missing order _for that type_ as being a validation error. I think it's different if you have a mix of ordered and non-ordered in the same inheritance/extension chain on the same entity type.